### PR TITLE
Remove unused/unnecessary features

### PR DESCRIPTION
**Pull request by clubby789**
~~The bulk of the actual code changes here is replacing try blocks with equivalent closures. I'm not entirely sure that's a good idea since it may have perf impact, happy to revert if that's the case/the change is unwanted.~~

I also removed a lot of `recursion_limit = "256"` since everything seems to build fine without that and most don't have any comment justifying it.

### CODE REVIEW

**Review by compiler-errors — COMMENTED** [review]
What's the motivation for removing try blocks? The rest of the changes seem fine, but there's no bad reason to keep "testing" try blocks in the compiler imo.

**Comment by clubby789** [comment]
Mostly just because it was a very easy change to make to cut down on unstable features - but yeah, I'll restore them

**Comment by WaffleLapkin** [comment]
Yeah, +1 to what @compiler-errors said — there isn't really a reason to avoid try blocks.

And while it's possible to remove their uses, it makes the code worse imo.

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs [hunk 1]
```diff
@@ -222,7 +222,7 @@ fn encode_mappings_for_function(
     function_coverage: &FunctionCoverage<'_>,
 ) -> Vec<u8> {
     let counter_regions = function_coverage.counter_regions();
-    if counter_regions.is_empty() {
+    if counter_regions.len() == 0 {
```
Comment (cjgillot) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:225]: Should we keep this one too?

File: compiler/rustc_mir_dataflow/src/framework/graphviz.rs [hunk 2]
```diff
@@ -438,8 +438,8 @@ where
         let after = next_in_dataflow_order(&mut diffs_after);
         let before = diffs_before.as_mut().map(next_in_dataflow_order);
 
-        assert!(diffs_after.is_empty());
-        assert!(diffs_before.as_ref().map_or(true, ExactSizeIterator::is_empty));
+        assert_eq!(diffs_after.len(), 0);
+        assert!(diffs_before.as_ref().map_or(true, |it| it.len() == 0));
```
Comment (cjgillot) [inline compiler/rustc_mir_dataflow/src/framework/graphviz.rs:442]: Should those be kept too?

**Review by compiler-errors — APPROVED** [review]
This is fine, though I guess I don't necessarily agree that we should be removing harmless libs features like `.get_or_insert_with_default()`.

**Comment by compiler-errors** [comment]
Should be fine tho

@bors r+

**Review by cjgillot — COMMENTED** [review]
Generally, should this PR be limited to removing the attributes? Testing library features is normal in the compiler.

**Comment by compiler-errors** [comment]
@bors r-

Yeah, let's just revert all the libs feature changes.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/119981#issuecomment-1891194288

**Comment by matthiaskrgr** [comment]
may need rebase
https://github.com/rust-lang/rust/pull/120087#issuecomment-1897951414
@bors r-

File: compiler/rustc_mir_dataflow/src/framework/cursor.rs [hunk 3]
```diff
@@ -185,8 +185,8 @@ where
         debug_assert_eq!(target.block, self.pos.block);
 
         let block_data = &self.body[target.block];
+        #[rustfmt::skip]
         let next_effect = if A::Direction::IS_FORWARD {
-            #[rustfmt::skip]
             self.pos.curr_effect_index.map_or_else(
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_dataflow/src/framework/cursor.rs:190]: ?

Comment (clubby789) [reply compiler/rustc_mir_dataflow/src/framework/cursor.rs:190]: That was to remove the `stmt_expr_attributes` for the crate. If I remove it entirely, I get
```diff
         let next_effect = if A::Direction::IS_FORWARD {
-            self.pos.curr_effect_index.map_or_else(
-                || Effect::Before.at_index(0),
-                EffectIndex::next_in_forward_order,
-            )
+            self.pos
+                .curr_effect_index
+                .map_or_else(|| Effect::Before.at_index(0), EffectIndex::next_in_forward_order)
```
which doesn't look bad imo

**Comment by clubby789** [comment]
@bors r-
Looks like  there was a conflict (the queue page says this isn't queued any more but no notification from Bors here)?

**Comment by clubby789** [comment]
Found another conflict while fixing that one so
@bors r=compiler-errors rollup=never

**Pull request merged**